src/rgw/rgw_crypt.cc

```diff
@@ -568,23 +568,22 @@ int RGWGetObj_BlockDecrypt::fixup_range(off_t& bl_ofs, off_t& bl_end) {
     off_t in_ofs = bl_ofs;
     off_t in_end = bl_end;
 
-    size_t i = 0;
-    while (i<parts_len.size() && (in_ofs > (off_t)parts_len[i])) {
-      in_ofs -= parts_len[i];
-      i++;
+    auto i = parts_len.cbegin();
+    while (i != parts_len.cend() && (in_ofs > (off_t)*i)) {
+      in_ofs -= *i;
+      ++i;
     }
     //in_ofs is inside block i
-    size_t j = 0;
-    while (j<parts_len.size() && (in_end > (off_t)parts_len[j])) {
-      in_end -= parts_len[j];
-      j++;
+    auto j = parts_len.cbegin();
+    while (j != parts_len.cend() && (in_end > (off_t)*j)) {
+      in_end -= *j;
+      ++j;
     }
     //in_end is inside block j
 
-    size_t rounded_end;
-    rounded_end = ( in_end & ~(block_size - 1) ) + (block_size - 1);
-    if (rounded_end + 1 >= parts_len[j]) {
-      rounded_end = parts_len[j] - 1;
+    size_t rounded_end = ( in_end & ~(block_size - 1) ) + (block_size - 1);
+    if (rounded_end > *j) {
+      rounded_end = *j;
```

Comment:
This still causes a offby1 error of sorts when you try a ranged get for exactly the size of multipart upload. I've a reproducible test case here: https://github.com/theanalyst/s3-tests/commit/29d6d0964451f47a172785ef9dc60c8ce44dff5c

Reply:
(passing a >= to in_end & in_ofs fixes that)

Reply:
the >= above makes sense - but my understanding is that the > prevented j from reaching the end() - so with >=, we now need to check before accessing `*j` here:
```c++
  if (j != parts_len.cend() && rounded_end > *j) {
```
do you buy that?

Reply:
I think this needs to be done, with a uniform multipart size this most likely cannot be hit, but if someone has a last part that exceeds the part_size then we'd likely reach here. 

Is there anyway we can set the parts_len vector in the unit tests without actually changing the public interface of the class? (the only method that sets the parts_len vector atm seems to be the read_manifest code)

Reply:
rounded_end needs to be set to j - 1 (@aclamk also has a similar commit fixing this)

Reply:
thanks, agreed - i removed my `rgw: off-by-one fix in RGWGetObj_BlockDecrypt::fixup_range()` commit and verified that your `>=` addresses the same issue and passes the tests in https://github.com/ceph/s3-tests/pull/266